xds/internal/xdsclient/xds.go

```diff
@@ -491,65 +518,72 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
+			case *v3routepb.RouteAction_Cluster:
+				route.WeightedClusters[a.Cluster] = WeightedCluster{Weight: 1}
+			case *v3routepb.RouteAction_WeightedClusters:
+				wcs := a.WeightedClusters
+				var totalWeight uint32
+				for _, c := range wcs.Clusters {
+					w := c.GetWeight().GetValue()
+					if w == 0 {
+						continue
 					}
-					wc.HTTPFilterConfigOverride = cfgs
+					wc := WeightedCluster{Weight: w}
+					if !v2 {
+						cfgs, err := processHTTPFilterOverrides(c.GetTypedPerFilterConfig())
+						if err != nil {
+							return nil, fmt.Errorf("route %+v, action %+v: %v", r, a, err)
+						}
+						wc.HTTPFilterConfigOverride = cfgs
+					}
+					route.WeightedClusters[c.GetName()] = wc
+					totalWeight += w
 				}
-				route.WeightedClusters[c.GetName()] = wc
-				totalWeight += w
-			}
-			// envoy xds doc
-			// default TotalWeight https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#envoy-v3-api-field-config-route-v3-weightedcluster-total-weight
-			wantTotalWeight := uint32(100)
-			if tw := wcs.GetTotalWeight(); tw != nil {
-				wantTotalWeight = tw.GetValue()
+				// envoy xds doc
+				// default TotalWeight https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#envoy-v3-api-field-config-route-v3-weightedcluster-total-weight
+				wantTotalWeight := uint32(100)
+				if tw := wcs.GetTotalWeight(); tw != nil {
+					wantTotalWeight = tw.GetValue()
+				}
+				if totalWeight != wantTotalWeight {
+					return nil, fmt.Errorf("route %+v, action %+v, weights of clusters do not add up to total total weight, got: %v, expected total weight from response: %v", r, a, totalWeight, wantTotalWeight)
+				}
+				if totalWeight == 0 {
+					return nil, fmt.Errorf("route %+v, action %+v, has no valid cluster in WeightedCluster action", r, a)
+				}
+			case *v3routepb.RouteAction_ClusterHeader:
+				continue
 			}
-			if totalWeight != wantTotalWeight {
-				return nil, fmt.Errorf("route %+v, action %+v, weights of clusters do not add up to total total weight, got: %v, expected total weight from response: %v", r, a, totalWeight, wantTotalWeight)
+
+			msd := action.GetMaxStreamDuration()
+			// Prefer grpc_timeout_header_max, if set.
+			dur := msd.GetGrpcTimeoutHeaderMax()
+			if dur == nil {
+				dur = msd.GetMaxStreamDuration()
 			}
-			if totalWeight == 0 {
-				return nil, fmt.Errorf("route %+v, action %+v, has no valid cluster in WeightedCluster action", r, a)
+			if dur != nil {
+				d := dur.AsDuration()
+				route.MaxStreamDuration = &d
 			}
-		case *v3routepb.RouteAction_ClusterHeader:
-			continue
-		}
-
-		msd := action.GetMaxStreamDuration()
-		// Prefer grpc_timeout_header_max, if set.
-		dur := msd.GetGrpcTimeoutHeaderMax()
-		if dur == nil {
-			dur = msd.GetMaxStreamDuration()
-		}
-		if dur != nil {
-			d := dur.AsDuration()
-			route.MaxStreamDuration = &d
+		case *v3routepb.Route_NonForwardingAction:
+			// Expected to be used on server side.
```

Comment:
The `action` is not persisted in the `Route` struct. Will this be done in a different PR?

Reply:
Oh, yeah you're right. I forgot to add it, will add a action struct. Also, I think we have a correctness issue here in regards to the default "route has invalid action". Eric's spec specifies this: "`Route.non_forwarding_action` is expected for all Routes used on server-side and
`Route.route` continues to be expected for all Routes used on client-side; a
Route with an inappropriate `action` causes RPCs matching that route to fail." Thus, it seems like we can have a route with an inappropriate action on either client or server, just the RPC's will fail. What do you think @menghanl

Reply:
Is `Route_NonForwardingAction` server side only? If so, client should NACK if it's present?

And similarly, is `Route_Route` client side only?

Reply:
No nack, just says it's "expected." Eric mentions that "A route with an inappropropriate action causes RPCs matching that route to fail" (my interpretation of this means that you can have any action, but if not not forwarding action on server or route on client fail RPC's.